pkg/internal/controller/controller.go

```diff
@@ -83,6 +83,22 @@ type Controller[request comparable] struct {
 	// startWatches maintains a list of sources, handlers, and predicates to start when the controller is started.
 	startWatches []source.TypedSource[request]
 
+	// didStartEventSourcesOnce is used to ensure that the event sources are only started once.
+	didStartEventSourcesOnce sync.Once
+
+	// ensureDidWarmupFinishChanInitializedOnce is used to ensure that the didWarmupFinishChan is
+	// initialized to a non-nil channel.
+	ensureDidWarmupFinishChanInitializedOnce sync.Once
+
+	// didWarmupFinish is closed when startEventSources returns. It is used to
+	// signal to WaitForWarmupComplete that the event sources have finished syncing.
+	didWarmupFinishChan chan struct{}
+
+	// didWarmupFinishSuccessfully is used to indicate whether the event sources have finished
+	// successfully. If true, the event sources have finished syncing without error. If false, the
+	// event sources have finished syncing but with error.
+	didWarmupFinishSuccessfully atomic.Bool
```

Comment:
Please make this a `chan error` and buffer it when initializing, so we can use it both to check if it finishes and if there was an issue and don't need the bool:
```suggestion
	didWarmupFinishChan chan error
```

Reply:
Resolving this one as the conversation moved on :)